Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestTranslator.mm

```diff
@@ -0,0 +1,108 @@
+
+@implementation _WKWebExtensionDeclarativeNetRequestTranslator
+
++ (NSArray<NSDictionary<NSString *, id> *> *)translateRules:(NSArray<NSArray<NSDictionary *> *> *)jsonObjects errorStrings:(NSArray **)outErrorStrings
+{
+    NSMutableArray<_WKWebExtensionDeclarativeNetRequestRule *> *allValidatedRules = [NSMutableArray array];
+    NSMutableArray<NSString *> *errorStrings = [NSMutableArray array];
+    NSUInteger totalErrorCount = 0;
+    for (NSArray *json in jsonObjects) {
+        for (NSDictionary *ruleJSON in json) {
+            NSString *errorString;
+            _WKWebExtensionDeclarativeNetRequestRule *rule = [[_WKWebExtensionDeclarativeNetRequestRule alloc] initWithDictionary:ruleJSON errorString:&errorString];
+
+            if (rule)
+                [allValidatedRules addObject:rule];
+            else if (errorString) {
+                totalErrorCount++;
+
+                if (errorStrings.count < maximumNumberOfDeclarativeNetRequestErrorsToSurface)
+                    [errorStrings addObject:errorString];
+            }
+        }
+    }
+
+    if (totalErrorCount > maximumNumberOfDeclarativeNetRequestErrorsToSurface)
+        [errorStrings addObject:@"Error limit hit. No longer omitting errors."];
+
+    if (outErrorStrings)
+        *outErrorStrings = [errorStrings copy];
+
+    allValidatedRules = [allValidatedRules sortedArrayUsingComparator:^NSComparisonResult(_WKWebExtensionDeclarativeNetRequestRule *a, _WKWebExtensionDeclarativeNetRequestRule *b) {
+        return [a compare:b];
+    }].mutableCopy;
+
+    NSMutableArray<NSDictionary<NSString *, id> *> *translatedRules = [NSMutableArray array];
+    for (_WKWebExtensionDeclarativeNetRequestRule *rule in allValidatedRules) {
+        NSArray<NSDictionary<NSString *, id> *> *translatedRule = rule.ruleInWebKitFormat;
+        [translatedRules addObjectsFromArray:translatedRule];
+    }
+
+    return translatedRules;
+}
+
++ (NSArray<NSArray<NSDictionary *> *> *)jsonObjectsFromData:(NSArray<NSData *> *)jsonDataArray errorStrings:(NSArray<NSString *> **)outErrorStrings
+{
+    NSMutableArray *allJSONObjects = [NSMutableArray array];
+    NSMutableArray<NSString *> *errors = [NSMutableArray array];
+    for (NSData *jsonData in jsonDataArray) {
+        NSError *error;
+        NSArray<NSDictionary *> *json = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&error];
+        if (json)
+            [allJSONObjects addObject:json];
+
+        if (error)
+            [errors addObject:error.userInfo[NSDebugDescriptionErrorKey]];
+    }
+
+    if (outErrorStrings)
+        *outErrorStrings = [errors copy];
+
+    return allJSONObjects;
+}
+
++ (NSString *)jsonStringForJSONObject:(id)jsonObject
+{
+    ASSERT([NSJSONSerialization isValidJSONObject:jsonObject]);
+    NSData *jsonAsData = [NSJSONSerialization dataWithJSONObject:jsonObject options:0 error:nil];
+    return [[NSString alloc] initWithData:jsonAsData encoding:NSUTF8StringEncoding];
+}
```

Comment:
There is a CocoaHelpers function that does this already.

Reply:
These aren't used yet - so I'll just get rid of them now and adopt the CocoaHelpers ones when adopting these classes.